File: iOS/starflower/starflower/GeoService.swift
//
//  GeoService.swift
//  starflower
//
//  Created by 양지성 on 6/19/26.
//

import Foundation

final class GeoService {
    static let shared = GeoService()
    private init() {}
    private let session = URLSession.shared

    private let placeTypes: Set<String> = [
        "city","town","village","hamlet","municipality",
        "suburb","neighbourhood","quarter","borough",
        "city_district","district","county","province",
        "state","region","administrative","island"
    ]

    private func isKorean(_ s: String) -> Bool {
        s.unicodeScalars.contains {
            ($0.value >= 0xAC00 && $0.value <= 0xD7A3) ||
            ($0.value >= 0x1100 && $0.value <= 0x11FF) ||
            ($0.value >= 0x3130 && $0.value <= 0x318F)
        }
    }

    func search(query: String) async -> [GeoResult] {
        let q = query.trimmingCharacters(in: .whitespaces)
        guard q.count >= 2 else { return [] }
        if let r = try? await searchNominatim(q), !r.isEmpty { return r }
        return (try? await searchOpenMeteo(q)) ?? []
    }

    // ── Nominatim ─────────────────────────────────────────
    private func searchNominatim(_ q: String) async throws -> [GeoResult] {
        if isKorean(q) {
            async let kr  = nominatimFetch(q, countryCode: "kr")
            async let all = nominatimFetch(q, countryCode: nil)
            let (krItems, allItems) = await ((try? kr) ?? [], (try? all) ?? [])
            let krIds = Set(krItems.map { $0.placeId })
            let merged = krItems + allItems.filter { !krIds.contains($0.placeId) }
            return parseItems(merged)
        } else {
            return parseItems(try await nominatimFetch(q, countryCode: nil))
        }
    }

    private func nominatimFetch(_ q: String, countryCode: String?) async throws -> [NominatimItem] {
        var c = URLComponents(string: "https://nominatim.openstreetmap.org/search")!
        var items: [URLQueryItem] = [
            .init(name: "q", value: q),
            .init(name: "format", value: "jsonv2"),
            .init(name: "addressdetails", value: "1"),
            .init(name: "accept-language", value: "ko"),
            .init(name: "limit", value: "10"),
        ]
        if let cc = countryCode { items.append(.init(name: "countrycodes", value: cc)) }
        c.queryItems = items

        var req = URLRequest(url: c.url!)
        req.setValue("ko", forHTTPHeaderField: "Accept-Language")
        req.setValue("Starflower/1.0 (stargazing app)", forHTTPHeaderField: "User-Agent")

        let (data, resp) = try await session.data(for: req)
        guard (resp as? HTTPURLResponse)?.statusCode == 200 else { throw GeoError.httpError }
        return try JSONDecoder().decode([NominatimItem].self, from: data)
    }

    private func parseItems(_ items: [NominatimItem]) -> [GeoResult] {
        let sorted = items.sorted { ($0.importance ?? 0) > ($1.importance ?? 0) }
        var seen = Set<String>()
        var out: [GeoResult] = []
        for it in sorted {
            let kind = it.addresstype ?? it.type ?? ""
            let isPlace = placeTypes.contains(kind)
                || it.category == "place" || it.category == "boundary"
            guard isPlace else { continue }
            guard let lat = Double(it.lat), let lon = Double(it.lon) else { continue }
            let a = it.address
            let name = it.name ?? a?.city ?? a?.town ?? a?.village
                ?? it.displayName.components(separatedBy: ",").first?
                    .trimmingCharacters(in: .whitespaces) ?? ""
            let admin1 = a?.state ?? a?.province
            let country = a?.country
            let key = "\(name)|\(admin1 ?? "")|\(country ?? "")"
            if seen.contains(key) { continue }
            seen.insert(key)
            out.append(GeoResult(id: it.placeId, name: name, admin1: admin1,
                                 country: country, latitude: lat, longitude: lon))
            if out.count >= 6 { break }
        }
        return out
    }

    // ── Open-Meteo 백업 ───────────────────────────────────
    private func searchOpenMeteo(_ q: String) async throws -> [GeoResult] {
        var c = URLComponents(string: "https://geocoding-api.open-meteo.com/v1/search")!
        c.queryItems = [
            .init(name: "name", value: q),
            .init(name: "count", value: "6"),
            .init(name: "language", value: "ko"),
            .init(name: "format", value: "json"),
        ]
        let (data, _) = try await session.data(from: c.url!)
        let r = try JSONDecoder().decode(OpenMeteoGeoResponse.self, from: data)
        return r.results?.map {
            GeoResult(id: $0.id, name: $0.name, admin1: $0.admin1,
                      country: $0.country, latitude: $0.latitude, longitude: $0.longitude)
        } ?? []
    }
}

private struct NominatimItem: Decodable {
    let placeId: Int
    let lat: String
    let lon: String
    let name: String?
    let displayName: String
    let addresstype: String?
    let type: String?
    let category: String?
    let importance: Double?
    let address: NominatimAddress?
    enum CodingKeys: String, CodingKey {
        case placeId = "place_id"
        case lat, lon, name
        case displayName = "display_name"
        case addresstype, type, category, importance, address
    }
}
private struct NominatimAddress: Decodable {
    let city: String?; let town: String?; let village: String?
    let county: String?; let state: String?; let province: String?; let country: String?
}
private struct OpenMeteoGeoResponse: Decodable { let results: [OpenMeteoGeoItem]? }
private struct OpenMeteoGeoItem: Decodable {
    let id: Int; let name: String; let latitude: Double; let longitude: Double
    let country: String?; let admin1: String?
}
enum GeoError: Error { case httpError }

File: Windows_Linux/starflower-desktop/src/main/kotlin/com/songsari/starflower/data/GeoService.kt
package com.songsari.starflower.data

import com.songsari.starflower.model.GeoResult
import com.squareup.moshi.Json
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.URLEncoder

/**
 * 지오코딩. Nominatim 을 우선 쓰고(한국어는 한국 우선 + 전세계 병렬),
 * 실패 시 Open-Meteo 백업. iOS/Android GeoService 와 동일한 동작.
 */
object GeoService {

    private val client = OkHttpClient()
    private val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()

    private val placeTypes = setOf(
        "city", "town", "village", "hamlet", "municipality",
        "suburb", "neighbourhood", "quarter", "borough",
        "city_district", "district", "county", "province",
        "state", "region", "administrative", "island"
    )

    private fun isKorean(s: String): Boolean = s.any {
        val v = it.code
        (v in 0xAC00..0xD7A3) || (v in 0x1100..0x11FF) || (v in 0x3130..0x318F)
    }

    suspend fun search(query: String): List<GeoResult> {
        val q = query.trim()
        if (q.length < 2) return emptyList()
        runCatching { searchNominatim(q) }.getOrNull()?.let { if (it.isNotEmpty()) return it }
        return runCatching { searchOpenMeteo(q) }.getOrNull() ?: emptyList()
    }

    // ── Nominatim ─────────────────────────────────────────
    private suspend fun searchNominatim(q: String): List<GeoResult> = coroutineScope {
        if (isKorean(q)) {
            val kr = async { runCatching { nominatimFetch(q, "kr") }.getOrDefault(emptyList()) }
            val all = async { runCatching { nominatimFetch(q, null) }.getOrDefault(emptyList()) }
            val krItems = kr.await()
            val allItems = all.await()
            val krIds = krItems.map { it.placeId }.toSet()
            parseItems(krItems + allItems.filter { it.placeId !in krIds })
        } else {
            parseItems(nominatimFetch(q, null))
        }
    }

    private suspend fun nominatimFetch(q: String, countryCode: String?): List<NominatimItem> =
        withContext(Dispatchers.IO) {
            val sb = StringBuilder("https://nominatim.openstreetmap.org/search")
            sb.append("?q=").append(URLEncoder.encode(q, "UTF-8"))
            sb.append("&format=jsonv2&addressdetails=1&accept-language=ko&limit=10")
            if (countryCode != null) sb.append("&countrycodes=").append(countryCode)

            val req = Request.Builder()
                .url(sb.toString())
                .header("Accept-Language", "ko")
                .header("User-Agent", "Starflower/1.0 (stargazing app)")
                .build()

            client.newCall(req).execute().use { resp ->
                if (!resp.isSuccessful) throw RuntimeException("http ${resp.code}")
                val body = resp.body?.string() ?: "[]"
                val type = Types.newParameterizedType(List::class.java, NominatimItem::class.java)
                val adapter = moshi.adapter<List<NominatimItem>>(type)
                adapter.fromJson(body) ?: emptyList()
            }
        }

    private fun parseItems(items: List<NominatimItem>): List<GeoResult> {
        val sorted = items.sortedByDescending { it.importance ?: 0.0 }
        val seen = HashSet<String>()
        val out = ArrayList<GeoResult>()
        for (it in sorted) {
            val kind = it.addresstype ?: it.type ?: ""
            val isPlace = kind in placeTypes || it.category == "place" || it.category == "boundary"
            if (!isPlace) continue
            val lat = it.lat.toDoubleOrNull() ?: continue
            val lon = it.lon.toDoubleOrNull() ?: continue
            val a = it.address
            val name = it.name
                ?: a?.city ?: a?.town ?: a?.village
                ?: it.displayName.split(",").firstOrNull()?.trim()
                ?: ""
            val admin1 = a?.state ?: a?.province
            val country = a?.country
            val key = "$name|${admin1 ?: ""}|${country ?: ""}"
            if (key in seen) continue
            seen.add(key)
            out.add(GeoResult(it.placeId, name, admin1, country, lat, lon))
            if (out.size >= 6) break
        }
        return out
    }

    // ── Open-Meteo 백업 ───────────────────────────────────
    private suspend fun searchOpenMeteo(q: String): List<GeoResult> =
        withContext(Dispatchers.IO) {
            val url = "https://geocoding-api.open-meteo.com/v1/search" +
                "?name=${URLEncoder.encode(q, "UTF-8")}&count=6&language=ko&format=json"
            val req = Request.Builder().url(url).build()
            client.newCall(req).execute().use { resp ->
                val body = resp.body?.string() ?: "{}"
                val parsed = moshi.adapter(OpenMeteoGeoResponse::class.java).fromJson(body)
                parsed?.results?.map {
                    GeoResult(it.id, it.name, it.admin1, it.country, it.latitude, it.longitude)
                } ?: emptyList()
            }
        }
}

data class NominatimItem(
    @Json(name = "place_id") val placeId: Int,
    val lat: String,
    val lon: String,
    val name: String?,
    @Json(name = "display_name") val displayName: String,
    val addresstype: String?,
    val type: String?,
    val category: String?,
    val importance: Double?,
    val address: NominatimAddress?,
)

data class NominatimAddress(
    val city: String?,
    val town: String?,
    val village: String?,
    val county: String?,
    val state: String?,
    val province: String?,
    val country: String?,
)

data class OpenMeteoGeoResponse(val results: List<OpenMeteoGeoItem>?)

data class OpenMeteoGeoItem(
    val id: Int,
    val name: String,
    val latitude: Double,
    val longitude: Double,
    val country: String?,
    val admin1: String?,
)
